flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java

```diff
@@ -249,6 +254,19 @@ public CoLocationConstraint getLocationConstraint() {
 		return locationConstraint;
 	}
 
+	public InputSplit getNextInputSplit(int index, String host) {
+		final int taskId = this.getParallelSubtaskIndex();
+		synchronized (this.inputSplits) {
+			if (index < this.inputSplits.size()) {
+				return this.inputSplits.get(index);
+			} else {
+				final InputSplit nextInputSplit = this.jobVertex.getSplitAssigner().getNextInputSplit(host, taskId);
```

Comment:
nit: When all input splits are exhausted for given vertex (nextInputSplit is null), you can just return null without adding an extra null element at the end of array list.

Reply:
put null here is to avoid recomputing, otherwise if there is another execution attempt pull getNextSplit we need to recompute.

Reply:
Ok, I think there shouldn't be too much re-computation overhead, since it happens once per attempt. I am fine with either way. Thx.